### PR TITLE
Account for different NSS error code in s390x

### PR DESCRIPTION
**Pull request by KostasTsiounis**
A difference in behaviour arises if the `SunPKCS11` provider, and the underlying `NSS` library, is used for verifying certificates in `s390x`. 

When the signature is incorrect and the verification of a certificate fails, the library returns a `CKR_DATA_LEN_RANGE` instead of a `CKR_SIGNATURE_LEN_RANGE` error.

The test is updated to account for this difference.

Fixes: https://github.com/eclipse-openj9/openj9/issues/18562

Signed-off by: Kostas Tsiounis [kostas.tsiounis@ibm.com](mailto:kostas.tsiounis@ibm.com)

### CODE REVIEW

**Comment by keithc-ca** [comment]
Would it not make more sense to fix `P11Signature` to anticipate `CKR_DATA_LEN_RANGE` than just ignore such test failures?

**Comment by keithc-ca** [comment]
That would also appear to have support from @pshipton - https://github.com/eclipse-openj9/openj9/issues/18562#issuecomment-1858179332.

**Comment by pshipton** [comment]
We wanted to purse a couple of things before making a non-OpenJDK "fix" to P11Signature. 1) check with the NSS open source project to see if a fix should be made there, 2) see about getting a fix acceptance at OpenJDK. If we proceed with a fix to P11Signature that hasn't been blessed upstream, we may be stuck supporting this change forever.

**Comment by keithc-ca** [comment]
Even if a fix to NSS is accepted, users may use older versions for some time, so accommodation for the distinct error code seems prudent.

**Comment by pshipton** [comment]
Perhaps, but so far no user has noticed the problem, and I don't believe it's a new issue.

**Comment by pshipton** [comment]
@KostasTsiounis for future reference, pls include the explanation why this is only needed in jdk17, not in jdk21, and not in jdk22+ (https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/724#issuecomment-1889730917 and fyi https://github.com/adoptium/aqa-tests/issues/4962). At the same time, pls add the explanation why it's not needed for jdk11 either.

**Comment by KostasTsiounis** [comment]
There is an additional issue that is covered up in other cases, however. When the call to `cancelOperation()` throws an exception, an error code provided by `NSS` is, also, attached. In most platforms, the error code is `CKR_DATA_LEN_RANGE`, whereas when the test is run on `s390x` a different error code is returned. In that case, the code is `CKR_SIGNATURE_LEN_RANGE`, which is what I am trying to cover with this PR.

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java [hunk 1]
```diff
@@ -55,6 +62,7 @@
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+
```
Comment (keithc-ca) [inline test/jdk/sun/security/pkcs11/ec/ReadCertificates.java:65]: Please don't make unnecessary changes relative to upstream.

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java [hunk 2]
```diff
@@ -182,6 +190,15 @@ public void main(Provider p) throws Exception {
                 }
             } catch (SignatureException | InvalidKeyException e) {
                 System.out.println("OK: " + e);
+            } catch (ProviderException pe) {
+                if (pe.getMessage().contains("cancel failed")
+                    && "s390x".equals(System.getProperty("os.arch"))
+                ) {
+                    System.out.println("NSS error code is different in s390x.");
```
Comment (keithc-ca) [inline test/jdk/sun/security/pkcs11/ec/ReadCertificates.java:197]: nit: "in" should be "on"

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java [hunk 3]
```diff
@@ -182,6 +190,15 @@ public void main(Provider p) throws Exception {
...
+                    pe.printStackTrace();
+                    return;
```
Comment (keithc-ca) [inline test/jdk/sun/security/pkcs11/ec/ReadCertificates.java:199]: This bypasses the code that prints "OK"; is that not necessary for the test to pass?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java (hunk 3 above)
Comment (keithc-ca) [reply test/jdk/sun/security/pkcs11/ec/ReadCertificates.java:199]: Maybe we want `break` instead of `return`?

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java (hunk 1 above)
Comment (KostasTsiounis) [reply test/jdk/sun/security/pkcs11/ec/ReadCertificates.java]: Reverted.

File: test/jdk/sun/security/pkcs11/ec/ReadCertificates.java (hunk 3 above)
Comment (KostasTsiounis) [reply test/jdk/sun/security/pkcs11/ec/ReadCertificates.java]: Yes, you are right. But we don't want to break. Rather move on to the next iteration. So, I updated accordingly.

**Comment by keithc-ca** [comment]
Jenkins compile zlinux jdk17

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
There was one test failure which apparently is rare:
```
23:31:33  java.lang.Exception: Client: Data length error - IF THIS FAILS, PLEASE REPORT THIS TO THE SECURITY TEAM.  WE HAVE BEEN UNABLE TO RELIABLY DUPLICATE.
23:31:33  	at SSLEngineBadBufferArrayAccess.runServerApplication(SSLEngineBadBufferArrayAccess.java:369)
23:31:33  	at SSLEngineBadBufferArrayAccess.runTest(SSLEngineBadBufferArrayAccess.java:247)
23:31:33  	at SSLEngineBadBufferArrayAccess.main(SSLEngineBadBufferArrayAccess.java:165)
23:31:33  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
23:31:33  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
23:31:33  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
23:31:33  	at java.base/java.lang.reflect.Method.invoke(Method.java:574)
23:31:33  	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
23:31:33  	at java.base/java.lang.Thread.run(Thread.java:857)
```

The specific test modified here passed in https://openj9-jenkins.osuosl.org/job/Grinder/3193.
